File: 송하연/11월 3주차/등굣길.java
import java.util.*;

// 행이나 열이 0이면 갈 수 있는 경로 1
// 현 위치의 경로 = 위 + 좌측 
// 웅덩이는 제외하기 = 0

class Solution {
    
    static final int MOD = 1000000007;
    
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;

        int[][] dp = new int[n][m];
        boolean[][] isPuddle = new boolean[n][m];
        
        for(int[] puddle:puddles){
            isPuddle[puddle[1]-1][puddle[0]-1]=true;
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                //시작점
                if(i==0 && j==0) {
                    dp[i][j]=1;
                    continue; 
                }
                
                //웅덩이
                if(isPuddle[i][j]) {
                    dp[i][j] = 0;
                    continue;
                }
                
                //상단 끝
                if(i==0 && j!=0) {
                    dp[i][j] = dp[i][j-1]%MOD;
                    continue;
                }
                
                //우측 끝
                if(j==0 && i!=0) {
                    dp[i][j] = dp[i-1][j]%MOD;
                    continue;
                }

                // 그 외
                dp[i][j] = dp[i-1][j]%MOD+dp[i][j-1]%MOD;
            }
        }
        
        return answer=dp[n-1][m-1]%MOD;
    }
    
}

File: 한승우/11월 4주차/등굣길.cpp
#include <string>
#include <vector>
using namespace std;

/*
거리 대신에 개수를 센다.
오른쪽, 아래로만 가기 때문에 무조건 최단 거리 보장인듯.
*/

const int DIV = 1000000007;

int solution(int m, int n, vector<vector<int>> puddles) {
    vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
    
    for(const auto& p: puddles) dp[p[1]][p[0]] = -1;
    
    dp[1][1] = 1;
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= m; j++){
            if(i == 1 && j == 1) continue;
            
            if(dp[i][j] == -1){
                dp[i][j] = 0;
                continue;
            }
            
            dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % DIV;
        }
    }
       
    return dp[n][m];
}